File: checkNIF.js
function checkNIF(nif) {
    nif = nif.toUpperCase().replace(/[_\W\s]+/g, '');
    if(/^(\d|[XYZ])\d{7}[A-Z]$/.test(nif)) {
        var num = nif.match(/\d+/);
        num = (nif[0]!='Z'? nif[0]!='Y'? 0: 1: 2)+num;
        if(nif[8]=='TRWAGMYFPDXBNJZSQVHLCKE'[num%23]) {
            return /^\d/.test(nif)? 'DNI': 'NIE';
        }
    }
    else if(/^[ABCDEFGHJKLMNPQRSUVW]\d{7}[\dA-J]$/.test(nif)) {
        for(var sum=0,i=1;i<8;++i) {
            var num = nif[i]<<i%2;
            sum += int(num/10)+num%10
        }
        var c = (10-su)%10;
        if( ((/[KLMNPQRSW]/.test(nif[0]) || (nif[1]+nif[2])=='00') && nif[8]=='JABCDEFGHI'[c]) ||
            (/[ABEH]/.test(nif[0]) && nif[8]==c) ||
            (/[CDFGJUV]/.test(nif[0]) && (nif[8]=='JABCDEFGHI'[c] || nif[8]==c)) ) {
            return /^[KLM]/.test(nif)? 'ESP': 'CIF';
        }
    }
    return false;
}

File: checkNIF.py
import re
def checkNIF(nif):
    nif = re.sub('[_\W\s]+','',nif.upper())
    if re.search('^(\d|[XYZ])\d{7}[A-Z]$',nif):
        num = re.findall('\d+',nif)[0]
        num = int(('2' if nif[0]=='Z' else '1' if nif[0]=='Y' else '0')+num)
        if nif[8]=='TRWAGMYFPDXBNJZSQVHLCKE'[num%23]:
            return 'DNI' if re.search('^\d',nif) else 'NIE'
    elif re.search('^[ABCDEFGHJKLMNPQRSUVW]\d{7}[\dA-J]$',nif):
        sum = 0;
        for i in range(1,8):
            num = int(nif[i])<<i%2
            sum += num//10+num%10
        c = (10-sum)%10
        if(((nif[0] in 'KLMNPQRSW' or nif[1:3]=='00') and nif[8]=='JABCDEFGHI'[c]) or
            (nif[0] in 'ABEH' and nif[8]==str(c)) or
            (nif[0] in 'CDFGJUV' and (nif[8]=='JABCDEFGHI'[c] or nif[8]==str(c))) ):
            return 'ESP' if re.search('^[KLM]',nif) else 'CIF'
    return False
